pkg/trace/event/extractor_single_span_rate.go

```diff
@@ -0,0 +1,36 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2016-present Datadog, Inc.
+
+package event
+
+import (
+	"github.com/DataDog/datadog-agent/pkg/remoteconfig/client/products/apmsampling"
+	"github.com/DataDog/datadog-agent/pkg/trace/pb"
+	"github.com/DataDog/datadog-agent/pkg/trace/sampler"
+	"github.com/DataDog/datadog-agent/pkg/trace/traceutil"
+)
+
+// singleSpanRateExtractor extracts spans that have been sampled using the single span sampling mechanism.
+type singleSpanRateExtractor struct{}
+
+func NewSingleSpanExtractor() Extractor {
+	return &singleSpanRateExtractor{}
+}
+
+func (e *singleSpanRateExtractor) Extract(s *pb.Span, _ sampler.SamplingPriority) (float64, bool) {
+	m, ok := traceutil.GetMetric(s, sampler.KeySpanSamplingMechanism)
+	if !ok || m != float64(apmsampling.SamplingMechanismSingleSpan) {
+		return 0, false
+	}
+	extractionRate, ok := s.Metrics[sampler.KeySpanSamplingRuleRate]
+	if !ok || extractionRate < 0 {
+		return 0, false
+	}
+	if extractionRate > 0 {
+		// If the trace has been manually sampled, we keep all matching spans
+		extractionRate = 1
+	}
```

Comment:
From the RFC:

> The Agent will be modified to recognize the _dd.span_sampling.mechanism tag and accordingly honor the associated "keep" sampling decision regardless of the decision made for the enclosing trace chunk.

It isn't clear what exactly the agent should look for in the `_dd.span_sampling.mechanism` tag.  Its presence?  Its value?

I think it's best to look for the presence of the tag only, so that we can introduce other mechanisms later without having to change the agent.

Then as for the rates, the tracer will only set the `_dd.span_sampling.*` tags when the span is kept by the rule, and so the value of the rate doesn't matter.  A negative rate would be strange, as would a zero rate, but I think it's better to leave the interpretation to the tracers only.

Consider something like this:
```Go
// Extract decides whether to extract a single span ingestion control event
// from the provided span having the specified priority. Extract returns a
// suggested extraction sample rate and a bool indicating whether an event was
// extracted. If the bool is false, then ignore the rate.
func (e *singleSpanRateExtractor) Extract(s *pb.Span, _ sampler.SamplingPriority) (float64, bool) {
        _, ok := traceutil.GetMetric(s, sampler.KeySpanSamplingMechanism)
        if ok {
                // If the tag is present, then the tracer wants us to keep the
                // span. The tracer already accounted for the rate.
                return 1, true
        }   

        return 0, false
}
```

If you do that, then you might be able to get rid of `sampler.KeySpanSamplingRuleRate` if it isn't used anywhere else.